File: hello-grpc-go/common/error_mapper.go
package common

import (
	"context"
	"fmt"
	"time"

	log "github.com/sirupsen/logrus"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"
)

// RetryConfig holds configuration for retry logic
type RetryConfig struct {
	MaxRetries   int
	InitialDelay time.Duration
	MaxDelay     time.Duration
	Multiplier   float64
}

// DefaultRetryConfig returns the default retry configuration
func DefaultRetryConfig() RetryConfig {
	return RetryConfig{
		MaxRetries:   3,
		InitialDelay: 2 * time.Second,
		MaxDelay:     30 * time.Second,
		Multiplier:   2.0,
	}
}

// MapGRPCError translates gRPC status codes to human-readable messages
func MapGRPCError(err error) string {
	if err == nil {
		return "Success"
	}

	st, ok := status.FromError(err)
	if !ok {
		return fmt.Sprintf("Unknown error: %v", err)
	}

	code := st.Code()
	message := st.Message()

	var description string
	switch code {
	case codes.OK:
		description = "Success"
	case codes.Canceled:
		description = "Operation cancelled"
	case codes.Unknown:
		description = "Unknown error"
	case codes.InvalidArgument:
		description = "Invalid request parameters"
	case codes.DeadlineExceeded:
		description = "Request timeout"
	case codes.NotFound:
		description = "Resource not found"
	case codes.AlreadyExists:
		description = "Resource already exists"
	case codes.PermissionDenied:
		description = "Permission denied"
	case codes.ResourceExhausted:
		description = "Resource exhausted"
	case codes.FailedPrecondition:
		description = "Precondition failed"
	case codes.Aborted:
		description = "Operation aborted"
	case codes.OutOfRange:
		description = "Out of range"
	case codes.Unimplemented:
		description = "Not implemented"
	case codes.Internal:
		description = "Internal server error"
	case codes.Unavailable:
		description = "Service unavailable"
	case codes.DataLoss:
		description = "Data loss"
	case codes.Unauthenticated:
		description = "Authentication required"
	default:
		description = "Unknown error code"
	}

	if message != "" {
		return fmt.Sprintf("%s: %s", description, message)
	}
	return description
}

// IsRetryableError determines if an error should be retried
func IsRetryableError(err error) bool {
	if err == nil {
		return false
	}

	st, ok := status.FromError(err)
	if !ok {
		return false
	}

	code := st.Code()
	switch code {
	case codes.Unavailable, codes.DeadlineExceeded, codes.ResourceExhausted, codes.Internal:
		return true
	default:
		return false
	}
}

// HandleRPCError logs and handles RPC errors with context
func HandleRPCError(err error, operation string, context map[string]interface{}) error {
	if err == nil {
		return nil
	}

	errorMsg := MapGRPCError(err)

	fields := log.Fields{
		"operation": operation,
		"error":     errorMsg,
	}

	for k, v := range context {
		fields[k] = v
	}

	if IsRetryableError(err) {
		log.WithFields(fields).Warn("Retryable error occurred")
	} else {
		log.WithFields(fields).Error("Non-retryable error occurred")
	}

	return err
}

// RetryWithBackoff executes a function with exponential backoff retry logic
func RetryWithBackoff(ctx context.Context, operation string, fn func() error, config RetryConfig) error {
	var lastErr error
	delay := config.InitialDelay

	for attempt := 0; attempt <= config.MaxRetries; attempt++ {
		if attempt > 0 {
			log.Infof("Retry attempt %d/%d for %s after %v", attempt, config.MaxRetries, operation, delay)

			select {
			case <-ctx.Done():
				return fmt.Errorf("operation cancelled: %w", ctx.Err())
			case <-time.After(delay):
				// Continue with retry
			}
		}

		lastErr = fn()
		if lastErr == nil {
			if attempt > 0 {
				log.Infof("Operation %s succeeded after %d attempts", operation, attempt+1)
			}
			return nil
		}

		if !IsRetryableError(lastErr) {
			log.Warnf("Non-retryable error for %s: %v", operation, MapGRPCError(lastErr))
			return lastErr
		}

		if attempt < config.MaxRetries {
			// Calculate next delay with exponential backoff
			delay = time.Duration(float64(delay) * config.Multiplier)
			if delay > config.MaxDelay {
				delay = config.MaxDelay
			}
		}
	}

	log.Errorf("Operation %s failed after %d attempts: %v", operation, config.MaxRetries+1, MapGRPCError(lastErr))
	return fmt.Errorf("max retries exceeded for %s: %w", operation, lastErr)
}

// LogError logs an error with request ID and operation context
func LogError(err error, requestID string, operation string) {
	if err == nil {
		return
	}

	errorMsg := MapGRPCError(err)
	log.WithFields(log.Fields{
		"request_id": requestID,
		"operation":  operation,
		"error":      errorMsg,
	}).Error("RPC error occurred")
}

// ToGrpcError converts an error to a gRPC status error with request ID
func ToGrpcError(err error, requestID string) error {
	if err == nil {
		return nil
	}

	// If it's already a gRPC status error, return it
	if _, ok := status.FromError(err); ok {
		return err
	}

	// Convert to gRPC error with request ID in metadata
	return status.Errorf(codes.Internal, "request_id=%s: %v", requestID, err)
}

File: hello-grpc-ts/common/errorMapper.ts
/**
 * Error mapper for translating gRPC status codes to human-readable messages
 * and implementing retry logic with exponential backoff.
 */

import * as grpc from '@grpc/grpc-js';
import { initializeLogging } from './loggingConfig';

const logger = initializeLogging('errorMapper', 'logs', false);

/**
 * Configuration for retry logic
 */
export class RetryConfig {
  constructor(
    public maxRetries: number = 3,
    public initialDelay: number = 2000, // milliseconds
    public maxDelay: number = 30000, // milliseconds
    public multiplier: number = 2.0
  ) {}

  static default(): RetryConfig {
    return new RetryConfig();
  }
}

/**
 * Type for gRPC errors
 */
export interface GrpcError extends Error {
  code?: grpc.status;
  details?: string;
}

/**
 * Maps gRPC status codes to human-readable error messages
 * @param error - The error to map
 * @returns Human-readable error message
 */
export function mapGrpcError(error: Error | GrpcError | null | undefined): string {
  if (!error) {
    return 'Success';
  }

  const grpcError = error as GrpcError;
  if (grpcError.code !== undefined) {
    const description = getStatusDescription(grpcError.code);
    const message = grpcError.details || grpcError.message || '';

    if (message) {
      return `${description}: ${message}`;
    }
    return description;
  }

  return `Unknown error: ${error.message || error}`;
}

/**
 * Gets human-readable description for a gRPC status code
 * @param code - The status code
 * @returns Human-readable description
 */
function getStatusDescription(code: grpc.status): string {
  const descriptions: Record<grpc.status, string> = {
    [grpc.status.OK]: 'Success',
    [grpc.status.CANCELLED]: 'Operation cancelled',
    [grpc.status.UNKNOWN]: 'Unknown error',
    [grpc.status.INVALID_ARGUMENT]: 'Invalid request parameters',
    [grpc.status.DEADLINE_EXCEEDED]: 'Request timeout',
    [grpc.status.NOT_FOUND]: 'Resource not found',
    [grpc.status.ALREADY_EXISTS]: 'Resource already exists',
    [grpc.status.PERMISSION_DENIED]: 'Permission denied',
    [grpc.status.RESOURCE_EXHAUSTED]: 'Resource exhausted',
    [grpc.status.FAILED_PRECONDITION]: 'Precondition failed',
    [grpc.status.ABORTED]: 'Operation aborted',
    [grpc.status.OUT_OF_RANGE]: 'Out of range',
    [grpc.status.UNIMPLEMENTED]: 'Not implemented',
    [grpc.status.INTERNAL]: 'Internal server error',
    [grpc.status.UNAVAILABLE]: 'Service unavailable',
    [grpc.status.DATA_LOSS]: 'Data loss',
    [grpc.status.UNAUTHENTICATED]: 'Authentication required',
  };

  return descriptions[code] || 'Unknown error code';
}

/**
 * Determines if an error should be retried
 * @param error - The error to check
 * @returns True if the error is retryable, false otherwise
 */
export function isRetryableError(error: Error | GrpcError | null | undefined): boolean {
  if (!error) {
    return false;
  }

  const grpcError = error as GrpcError;
  if (grpcError.code === undefined) {
    return false;
  }

  const retryableCodes = [
    grpc.status.UNAVAILABLE,
    grpc.status.DEADLINE_EXCEEDED,
    grpc.status.RESOURCE_EXHAUSTED,
    grpc.status.INTERNAL,
  ];

  return retryableCodes.includes(grpcError.code);
}

/**
 * Handles RPC errors with logging and context
 * @param error - The error that occurred
 * @param operation - The operation name
 * @param context - Additional context information
 */
export function handleRpcError(
  error: Error | GrpcError | null | undefined,
  operation: string,
  context: Record<string, any> = {}
): void {
  if (!error) {
    return;
  }

  const errorMsg = mapGrpcError(error);
  const logContext = {
    ...context,
    operation,
    error: errorMsg,
  };

  if (isRetryableError(error)) {
    logger.warn(`Retryable error occurred: ${JSON.stringify(logContext)}`);
  } else {
    logger.error(`Non-retryable error occurred: ${JSON.stringify(logContext)}`);
  }
}

/**
 * Sleeps for the specified duration
 * @param ms - Milliseconds to sleep
 */
function sleep(ms: number): Promise<void> {
  return new Promise(resolve => setTimeout(resolve, ms));
}

/**
 * Executes a function with exponential backoff retry logic
 * @param operation - The operation name for logging
 * @param func - The function to execute (can return a Promise)
 * @param config - The retry configuration
 * @returns The result of the function
 * @throws Error if all retries are exhausted
 */
export async function retryWithBackoff<T>(
  operation: string,
  func: () => T | Promise<T>,
  config: RetryConfig | null = null
): Promise<T> {
  if (!config) {
    config = RetryConfig.default();
  }

  let lastError: Error | null = null;
  let delay = config.initialDelay;

  for (let attempt = 0; attempt <= config.maxRetries; attempt++) {
    if (attempt > 0) {
      logger.info(
        `Retry attempt ${attempt}/${config.maxRetries} for ${operation} after ${delay}ms`
      );
      await sleep(delay);
    }

    try {
      const result = await Promise.resolve(func());
      if (attempt > 0) {
        logger.info(`Operation ${operation} succeeded after ${attempt + 1} attempts`);
      }
      return result;
    } catch (error) {
      lastError = error as Error;

      if (!isRetryableError(lastError)) {
        logger.warn(`Non-retryable error for ${operation}: ${mapGrpcError(lastError)}`);
        throw lastError;
      }

      if (attempt < config.maxRetries) {
        // Calculate next delay with exponential backoff
        delay = Math.min(delay * config.multiplier, config.maxDelay);
      }
    }
  }

  logger.error(
    `Operation ${operation} failed after ${config.maxRetries + 1} attempts: ${mapGrpcError(lastError)}`
  );
  const error = new Error(`Max retries exceeded for ${operation}`);
  (error as any).cause = lastError;
  throw error;
}
